### PR TITLE
move common logic of highestSupportedVersion to util package

### LINKED ISSUE #113831: dynamic resource allocation: move common logic of highestSupportedVersion to util package
See comment for details:
https://github.com/cncf-tags/container-device-interface/pull/93

### PR DESCRIPTION
**Pull request by tmilox**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

If applied this commit will move the common logic of the `highestSupportedVersion` to  the util package.

#### Which issue(s) this PR fixes:

Fixes #113831

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by apelisse** [comment]
I don't have a strong opinion, that looks fine to me.
/approve

**Comment by bart0sh** [comment]
/assign @mrunalp @derekwaynecarr @dchen1107 
for the final approval

**Comment by klueska** [comment]
CCing sig-storage-approvers for final approval
/cc @gnufied @safrane @msau42 @saad-ali @xing-yang

File: pkg/volume/csi/csi_plugin.go [hunk 1]
```diff
@@ -858,44 +858,6 @@ func unregisterDriver(driverName string) error {
...
-		currentHighestVer, err := utilversion.ParseGeneric(versions[i])
-		if err != nil {
-			theErr = err
-			continue
-		}
-		if currentHighestVer.Major() > 1 {
-			// CSI currently only has version 0.x and 1.x (see https://github.com/container-storage-interface/spec/releases).
-			// Therefore any driver claiming version 2.x+ is ignored as an unsupported versions.
-			// Future 1.x versions of CSI are supposed to be backwards compatible so this version of Kubernetes will work with any 1.x driver
-			// (or 0.x), but it may not work with 2.x drivers (because 2.x does not have to be backwards compatible with 1.x).
```
Comment (xing-yang) [inline pkg/volume/csi/csi_plugin.go:879]: Some of these comments should be moved to line 160 where the function is called in csi_plugin now that this function itself is moved to the utilVersion.

File: staging/src/k8s.io/apimachinery/pkg/util/version/version.go [hunk 2]
```diff
@@ -85,6 +86,46 @@ func parse(str string, semver bool) (*Version, error) {
 	return v, nil
 }
 
+// HighestSupportedVersion returns the highest supported version
+func HighestSupportedVersion(versions []string) (*Version, error) {
```
Comment (xing-yang) [inline staging/src/k8s.io/apimachinery/pkg/util/version/version.go:90]: Add the following comments:
This function assumes that the highest supported version must be v1.x.

File: staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go [hunk 3]
```diff
@@ -346,3 +346,109 @@ func TestComponents(t *testing.T) {
...
+			shouldFail: true,
+		},
+		{
+			versions:   []string{"0.2.0"},
+			shouldFail: true,
+		},
+		{
+			versions:                        []string{"1.0.0"},
+			expectedHighestSupportedVersion: "1.0.0",
+			shouldFail:                      false,
```
[the remark this answers is not in the document]
Comment (TommyStarK) [reply staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go:372]: no leading `v`

File: staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go [hunk 4]
```diff
@@ -346,3 +346,109 @@ func TestComponents(t *testing.T) {
...
+		},
+		{
+			versions:   []string{"v0.3.0"},
+			shouldFail: true,
```
[the remark this answers is not in the document]
Comment (TommyStarK) [reply staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go:376]: there is a leading `v`

File: staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go [hunk 5]
```diff
@@ -346,3 +346,109 @@ func TestComponents(t *testing.T) {
...
+		},
+		{
+			versions:   []string{"0.2.0"},
+			shouldFail: true,
```
[the remark this answers is not in the document]
Comment (TommyStarK) [reply staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go:380]: ack. Tests have been moved so it is already the case on master

File: staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (TommyStarK) [reply staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go:372]: no leading `v`

File: staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (TommyStarK) [reply staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go:376]: there is a leading `v`

File: staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (TommyStarK) [reply staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go:380]: ack. However tests have been only moved. This is already the case on master

File: staging/src/k8s.io/apimachinery/pkg/util/version/version.go [hunk 6]
```diff
@@ -85,6 +86,46 @@ func parse(str string, semver bool) (*Version, error) {
...
+		}
+
+		if highestSupportedVersion == nil || highestSupportedVersion.LessThan(currentHighestVer) {
+			highestSupportedVersion = currentHighestVer
+		}
+	}
+
+	if highestSupportedVersion == nil {
+		return nil, fmt.Errorf(
+			"could not find a highest supported version from versions (%v) reported by: %+v",
```
[the remark this answers is not in the document]
Comment (TommyStarK) [reply staging/src/k8s.io/apimachinery/pkg/util/version/version.go:118]: ack

File: pkg/volume/csi/csi_plugin.go (hunk 1 above)
Comment (TommyStarK) [reply pkg/volume/csi/csi_plugin.go:879]: ack

File: pkg/volume/csi/csi_plugin.go [hunk 7]
```diff
@@ -858,44 +858,6 @@ func unregisterDriver(driverName string) error {
...
-		}
-	}
-
-	if highestSupportedVersion == nil {
-		return nil, fmt.Errorf("could not find a highest supported version from versions (%v) reported by this driver: %v", versions, theErr)
-	}
-
-	if highestSupportedVersion.Major() != 1 {
-		// CSI v0.x is no longer supported as of Kubernetes v1.17 in
-		// accordance with deprecation policy set out in Kubernetes v1.13
```
[the remark this answers is not in the document]
Comment (TommyStarK) [reply pkg/volume/csi/csi_plugin.go:893]: ack

File: staging/src/k8s.io/apimachinery/pkg/util/version/version.go (hunk 2 above)
Comment (TommyStarK) [reply staging/src/k8s.io/apimachinery/pkg/util/version/version.go:90]: ack

File: staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (xing-yang) [reply staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go:376]: I see

File: staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (xing-yang) [reply staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go:372]: Ok

File: staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (xing-yang) [reply staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go:380]: I guess you can add a leading "v" in one of the tests.

[the remark this answers is not in the document]
Comment (TommyStarK) [reply staging/src/k8s.io/apimachinery/pkg/util/version/version_test.go:380]: done in 42356bfbb3e

File: staging/src/k8s.io/apimachinery/pkg/util/version/version.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (TommyStarK) [reply staging/src/k8s.io/apimachinery/pkg/util/version/version.go]: done in 42356bfbb3e

File: pkg/volume/csi/csi_plugin.go (hunk 1 above)
Comment (TommyStarK) [reply pkg/volume/csi/csi_plugin.go:879]: done in 42356bfbb3e

File: staging/src/k8s.io/apimachinery/pkg/util/version/version.go (hunk 2 above)
Comment (TommyStarK) [reply staging/src/k8s.io/apimachinery/pkg/util/version/version.go:91]: done in 42356bfbb3e

File: pkg/volume/csi/csi_plugin.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (TommyStarK) [reply pkg/volume/csi/csi_plugin.go:893]: done in 42356bfbb3e

**Comment by TommyStarK** [comment]
@xing-yang PTAL

**Pull request merged**